File: grains/grains.clj
(ns grains)

(defn- expt [b p]
  (loop [b (bigint b), p p, z 1]
    (cond
      (zero? p) z
      (even? p) (recur (* b b) (/ p 2) z)
      :else     (recur b (dec p) (* b z)))))

(def square (comp (partial expt 2) dec))

(def total (constantly (dec (expt 2 64))))

File: grains/grains.erl
-module(grains).
-export([square/1, total/0]).

square(N) -> trunc(math:pow(2, N-1)).

total() -> square(65) - 1.
%% total() -> trunc(math:pow(2, 64)) - 1.
%% total() -> 18446744073709551615.
